src/random.cr

```diff
@@ -1,34 +1,179 @@
-lib C
-  RAND_MAX = 2147483647
+class Random
 
-  fun srand(seed : UInt32)
-  fun rand() : Int32
-end
+  class MT19937
+    N = 624
+    M = 397
+    MATRIX_A = 0x9908b0dfu32
+    UPPER_MASK = 0x80000000u32
+    LOWER_MASK = 0x7fffffffu32
 
-C.srand(Intrinsics.read_cycle_counter.to_u32)
+    def initialize(seeds = Array(UInt32).new(4){ Intrinsics.read_cycle_counter.to_u32 })
+      @mt = StaticArray(UInt32, 624).new(0u32)
+      @mti = N + 1
+      init_by_array(seeds)
+    end
 
-def srand(x)
-  C.srand(x.to_u32)
+    private def init_genrand(seed)
+      @mt[0] = seed & 0xffffffffu32
+      @mti = 1
+      while @mti < N
+        @mt[@mti] = (1812433253u32 * (@mt[@mti-1] ^ (@mt[@mti-1] >> 30)) + @mti) & 0xffffffffu32
+        @mti += 1
+      end
+    end
+
+    private def init_by_array(init_keys)
+      key_len = init_keys.size
+      init_genrand 19650218u32
+
+      i = 1
+      j = 0
+      k = if N > key_len
+            N
+          else
+            key_len
+          end
+
+      while k > 0
+        @mt[i] = (@mt[i] ^ ((@mt[i-1] ^ (@mt[i-1] >> 30)) * 1664525u32)) + init_keys[j] + j
+
+        i += 1
+        j += 1
+
+        if i >= N
+          @mt[0] = @mt[N-1]
+          i = 1
+        end
+
+        if j >= key_len
+          j = 0
+        end
+
+        k -= 1
+      end
+
+      k = N - 1
+
+      while k > 0
+        @mt[i] = (@mt[i] ^ ((@mt[i-1] ^ (@mt[i-1] >> 30)) * 1566083941u32)) - i
+        i += 1
+
+        if i >= N
+          @mt[0] = @mt[N-1]
+          i = 1
+        end
+
+        k -= 1
+      end
+
+      # Use to_i because substituting 0x80000000 causes SEGV
+      @mt[0] = 0x80000000u32
+    end
+
+    def next_number()
+      mag01 = [0, MATRIX_A]
```

Comment:
I did this small benchmark:

```crystal
m = Random::MT19937.new [0x123, 0x234, 0x345, 0x456]

time = Time.new
a = 0_u32
10_000_000.times { a += m.next_number }
puts a
puts Time.new - time
```

It takes 2.16 seconds on my machine. But, if I change `mag01` to be `mag01 = {0, MATRIX_A}` (a tuple, no memory is allocated) the time goes down to 0.072 seconds! Then, if instead of using a tuple we just do `(y % 2 == 0 ? 0 : MATRIX_A)` the time goes down to 0.037 seconds. I thought the compiler (or LLVM ;-)) would optimize it, but no, so I prefer for now a slightly longer version but more performant.

Is it OK for you to change it to the later version?